File: day-01-variables-and-types/python/exercise01.py
"""
============================================================================
DAY 01 EXERCISE (Python) — Price Calculator
============================================================================
Apply today's concepts: dynamic typing, mutability, == vs is, function args.
Fill in every TODO, then run:  python3 exercise01.py
Target time: 15–20 minutes.
============================================================================
"""


def main():
    # ------------------------------------------------------------------
    # TASK 1: Create these variables and print each with its type
    # using f-strings and type().__name__:
    #   item_name = "Mechanical Keyboard"
    #   price     = 79.99
    #   quantity  = 3
    #   in_stock  = True
    # Example output line: item_name = 'Mechanical Keyboard' (str)
    # ------------------------------------------------------------------
    # TODO 1


    # ------------------------------------------------------------------
    # TASK 2: Compute line_total = price * quantity and print it.
    # Notice the floating point result! Then fix the display by printing
    # it rounded to 2 decimals with an f-string format spec.
    # Interview nugget: why is 79.99 * 3 not exactly 239.97?
    # (Answer in a comment — keyword: binary floating point / IEEE 754)
    # ------------------------------------------------------------------
    # TODO 2


    # ------------------------------------------------------------------
    # TASK 3: Build the label "Mechanical Keyboard x3 = $239.97"
    # two ways:
    #   a) f-string
    #   b) str.format()
    # Print both.
    # ------------------------------------------------------------------
    # TODO 3


    # ------------------------------------------------------------------
    # TASK 4: Fix the bug. apply_discount_buggy is SUPPOSED to return a
    # new price with 10% off. It doesn't. Predict the output first,
    # then fix it.
    # ------------------------------------------------------------------
    result = apply_discount_buggy(100.0)
    print("buggy discount result:", result)   # prediction?

    # TODO 4: fix apply_discount_buggy (or write apply_discount) so the
    # printed result is 90.0


    # ------------------------------------------------------------------
    # TASK 5: Lists are mutable and shared. Write apply_tax(prices) that
    # multiplies every element IN PLACE by 1.08 (rounded to 2 decimals).
    # Print the cart before and after.
    # Then answer in a comment: how would you write it to NOT mutate the
    # caller's list? (You don't have to code it — just say how.)
    # ------------------------------------------------------------------
    cart = [10.0, 20.0, 30.0]
    # TODO 5


    # ------------------------------------------------------------------
    # TASK 6: The mutable-default trap. Fix add_to_cart so that two
    # separate calls without a cart argument do NOT share a list.
    # ------------------------------------------------------------------
    # print(add_to_cart("apple"))    # expect ['apple']
    # print(add_to_cart("banana"))   # expect ['banana']  (not ['apple', 'banana'])
    # TODO 6


def apply_discount_buggy(price):
    price = price * 0.90   # something is missing — what?


def add_to_cart(item, cart=[]):
    cart.append(item)
    return cart


if __name__ == "__main__":
    main()

File: day-01-variables-and-types/java/Exercise01.java
/**
 * ============================================================================
 * DAY 01 EXERCISE (Java) — Price Calculator
 * ============================================================================
 * Apply today's concepts: primitives vs references, immutability, pass-by-value.
 * Fill in every TODO, then run:  javac Exercise01.java && java Exercise01
 * Target time: 20 minutes.
 * ============================================================================
 */
public class Exercise01 {

    public static void main(String[] args) {
        // ---------------------------------------------------------------
        // TASK 1: Declare variables with the CORRECT types.
        //   - itemName: the text "Mechanical Keyboard"
        //   - price: 79.99 (which primitive? hint: not float — interviews
        //     expect you to know float's precision issues)
        //   - quantity: 3
        //   - inStock: true
        // Print each with its value.
        // ---------------------------------------------------------------
        // TODO 1


        // ---------------------------------------------------------------
        // TASK 2: Compute the line total (price * quantity) and store it
        // in a variable called lineTotal. Print it.
        // Question to answer in a comment: what TYPE is the result, and why?
        // ---------------------------------------------------------------
        // TODO 2


        // ---------------------------------------------------------------
        // TASK 3: Build a label "Mechanical Keyboard x3 = $239.97"
        // WITHOUT using + concatenation in a loop — this is a single
        // string, so + is fine here, BUT do it two ways:
        //   a) with + concatenation
        //   b) with String.format (interviewers love seeing you know it)
        // Print both.
        // ---------------------------------------------------------------
        // TODO 3


        // ---------------------------------------------------------------
        // TASK 4: Fix the bug. This method is SUPPOSED to apply a 10%
        // discount to the price. Predict what it prints BEFORE running,
        // then fix it so the discount actually applies.
        // ---------------------------------------------------------------
        double price4 = 100.0;
        applyDiscountBuggy(price4);
        System.out.println("price after buggy discount: " + price4); // prediction?

        // TODO 4: write applyDiscountFixed() that RETURNS the discounted
        // price, call it, and print the result (should be 90.0)


        // ---------------------------------------------------------------
        // TASK 5: Arrays share references. Given cart below, write a method
        // applyTax(double[] prices) that multiplies EVERY element in place
        // by 1.08. Call it, print the cart before and after.
        // ---------------------------------------------------------------
        double[] cart = {10.0, 20.0, 30.0};
        // TODO 5


        // ---------------------------------------------------------------
        // STRETCH (optional, if time remains):
        // Write a method isSameProduct(String a, String b) that compares
        // two product names correctly. Test it with:
        //   isSameProduct("Mouse", new String("Mouse"))  -> must be true
        // Explain in a comment why the naive version fails.
        // ---------------------------------------------------------------
        // TODO stretch
    }

    static void applyDiscountBuggy(double price) {
        price = price * 0.90;   // why doesn't this work? (answer in a comment)
    }
}
